### PR TITLE
fix: service_name and endpoint not found handling for watchers

### LINKED ISSUE #1198: Security-critical watcher should handle unsupported API endpoints gracefully
Security-critical watcher should return the proper JSON-formatted error response when a request on the informational API is made to the security-critical watcher.

## Actual behaviour

Security-critical watcher is throwing this error when request is received on `/transaction.all`:

```elixir
UndefinedFunctionError: function OMG.WatcherInfo.API.Transaction.get_transactions/1 is undefined (module OMG.WatcherInfo.API.Transaction is not available)
  Module "Elixir.OMG.WatcherInfo.API.Transaction", in OMG.WatcherInfo.API.Transaction.get_transactions/1
  File "lib/web/controllers/transaction.ex", line 43, in OMG.WatcherRPC.Web.Controller.Transaction.get_transactions/2
  File "lib/web/controllers/transaction.ex", line 15, in OMG.WatcherRPC.Web.Controller.Transaction.action/2
  File "lib/web/controllers/transaction.ex", line 15, in OMG.WatcherRPC.Web.Controller.Transaction.phoenix_controller_pipeline/2
  File "lib/web/endpoint.ex", line 15, in OMG.WatcherRPC.Web.Endpoint.instrument/4
  File "lib/spandex_phoenix.ex", line 156, in OMG.WatcherRPC.Web.Controller.Transaction.call/2
  File "lib/phoenix/router.ex", line 280, in Phoenix.Router.__call__/2
  File "lib/web/endpoint.ex", line 15, in OMG.WatcherRPC.Web.Endpoint.plug_builder_call/2
```

## Expected behaviour

Watcher should return the proper error response, e.g.:

```json
{
  "data": {
    "code": "operation:not_found",
    "description": "Operation cannot be found. Check request URL.",
    "object": "error"
  },
  "service_name": "watcher",
  "success": false,
  "version": "0.3.0+94d24f0"
}
```

### LINKED ISSUE #1201: API to return service_name: "watcher" or "watcher_info" based on the started mode
As an API user I want to check the response's `"service_name"` to know whether I'm talking to the security-critical watcher API or the informational one.

Original conversation: https://github.com/omisego/elixir-omg/pull/1193#issuecomment-564222621

## Actual behaviour

Both `watcher` and `watcher_info` return `"service_name": "watcher"`.

```json
{
  "data": {
    "some_key": "some value"
  },
  "service_name": "watcher",
  "success": true,
  "version": "0.3.0+94d24f0"
}
```

## Expected behaviour

- `watcher` return `"service_name": "watcher"`
- `watcher_info` return `"service_name": "watcher_info"`
- The value is determined by the start mode, not where the endpoint belongs now. This is because `watcher_info`'s API specs will eventually include all security-critical endpoints.

### PR DESCRIPTION
**Pull request by unnawut**
Closes #1198 
Closes #1201 

## Overview

This PR builds on top of @pdobacz's work in #1193 and add two extra behaviours:

1. `watcher` and `watcher_info` return API responses with `service_name: watcher` and `service_name: watcher_info` respectively

2. Calling informational API endpoints to the `watcher` returns a properly handled error `operation:not_found`

## Changes

**Changes already present in #1193 (thanks to @pdobacz):**
- Add `version` and `service_name` (service name doesn't differentiate `watcher` and `watcher_info` yet)
- More robust `health_test` (allows multiple alarms)

**New changes:**

- Removed `conn.assigns.app_infos` because it's not used along the pipeline, only at the end
- Remove unused `${RELEASE_ROOT_DIR}...` arguments from all config providers
- Refactor child chain controllers to use `api_response()` consistently, instead of some doing their own view rendering and some using `api_response()`
- Refactor fallback controllers so error's json rendering is consistently done at view level, instead of some doing in the controller and some in the view
- Add `OMG.WatcherRPC.Web.Plugs.SupportedWatcherModes` that checks if its router scope should serve the currently running watcher mode
- Consistently add `version` and `service_name` at view level instead of some doing at controllers, some at views and some in other helper functions
- Unit tests

## Discussion notes

The existing code is single `OMG.WatcherRPC` routing through `OMG.Watcher` and `OMG.WatcherInfo` which makes this PR a little tricky. A few options:

1. Evaluate the routing at compile time based on distillery release targets
    - Evaluates at compile-time, more performant
    - No easy way to pass the release target to the router evaluation other than putting `Code.ensure_loaded?/1` directly in the router
    - More ideal way is to have two separate RPC subapps for watcher and watcher_info. But since there's a lot of shared validators, controllers and views, this is a big effort
2. Add a higher-order router that does switches on/off the routing to `Router.InfoAPI` and `Router.SecurityAPI`
    - Evaluates at run-time, less performant
    - Complicates the routing quite a bit. We would need 3 routers for the watcher
3. Add a plug that evaluates the start mode before it hits the router (thanks @InoMurko for the suggestion)
    - Evaluates at run-time, less performant
    - Scope ordering in the router is important since the routes are evaluated sequentially

The plug option is chosen for this PR for its simplicity and relatively quick turnaround time.

## Testing

- `mix test test/omg_watcher_rpc/web/plugs/supported_watcher_modes_test.exs`
- `mix test test/omg_watcher_rpc/web/response_test.exs`
- Use the API:

    ```shell
    # Calling /account.get_balance on watcher_info (port 7434) should return the handled request
    $ curl http://localhost:7534/account.get_balance \
    -H "Content-Type: application/json" \
    -d '{
      "address": "0x0433420dee34412b5bf1e29fbf988ad037cc5db7"
    }' \
    -v -w "\n" | jq

    {
      "data": [],
      "service_name": "watcher_info",
      "success": true,
      "version": "0.3.0+df1b4bb"
    }
    ```

    ```shell
    # Calling /account.get_balance on watcher (port 7434) should return operation:not_found
    $ curl http://localhost:7434/account.get_balance \
    -H "Content-Type: application/json" \
    -d '{
      "address": "0xc0f780dfc35075979b0def588d999225b7ecc56f"
    }' \
    -v -w "\n" | jq

    {
      "data": {
        "code": "operation:not_found",
        "description": "Operation cannot be found. Check request URL.",
        "object": "error"
      },
      "service_name": "watcher",
      "success": false,
      "version": "0.3.0+df1b4bb"
    }
    ```

### CODE REVIEW

**Comment by unnawut** [comment]
Swagger specs update coming in a bit...

File: apps/omg_watcher_rpc/lib/web/plugs/supported_watcher_modes.ex [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+defmodule OMG.WatcherRPC.Web.Plugs.SupportedWatcherModes do
+  @moduledoc """
+  A plug that serves the endpoints in scope only when the running watcher's mode matches
```
Comment (pdobacz) [inline apps/omg_watcher_rpc/lib/web/plugs/supported_watcher_modes.ex]: I find it hard to get through this sentence. How about "Ensures that unsupported endpoints return an appropriate error, depending on watcher's mode"?

File: apps/omg_watcher_rpc/lib/web/plugs/supported_watcher_modes.ex [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+  the plug's supported watcher modes.
+  """
+  alias OMG.WatcherRPC.Web.Controller
+
+  @behaviour Plug
+  @app :omg_watcher_rpc
+
+  @spec init([atom()]) :: [atom()]
+  def init(supported), do: supported
```
Comment (pdobacz) [inline apps/omg_watcher_rpc/lib/web/plugs/supported_watcher_modes.ex]: in this and `call/2` let's call `supported_modes` and `supported_mode`

File: apps/omg_watcher_rpc/test/omg_watcher_rpc/web/response_test.exs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+    on_exit(fn ->
+      _ = Application.put_env(@app, :api_mode, original_mode)
+    end)
+  end
+
+  describe "add_app_infos/1" do
+    test "appends the given map with a service_name and version" do
+      :ok = Application.put_env(@app, :api_mode, :watcher)
+      assert %{foo: "bar", service_name: _, version: _} = Response.add_app_infos(%{foo: "bar"})
```
Comment (pdobacz) [inline apps/omg_watcher_rpc/test/omg_watcher_rpc/web/response_test.exs]: how about testing if the `version` field holds in fact a version as in [here](https://github.com/omisego/elixir-omg/pull/1266/commits/19063bd63ace82b248f9794a10c319b4361405f2#diff-3c6511e75122d06d1b190c6548d3af09R30)

> `assert {:ok, _} = Version.parse(conn.assigns.app_infos.version)`

?

Comment (unnawut) [reply apps/omg_watcher_rpc/test/omg_watcher_rpc/web/response_test.exs]: Ah I missed this. Added in 5b3ef09 and b88a2c5.

File: apps/omg_watcher_rpc/lib/web/plugs/supported_watcher_modes.ex (hunk 1 above)
Comment (unnawut) [reply apps/omg_watcher_rpc/lib/web/plugs/supported_watcher_modes.ex]: 1aa0f3d

File: apps/omg_watcher_rpc/lib/web/plugs/supported_watcher_modes.ex (hunk 2 above)
Comment (unnawut) [reply apps/omg_watcher_rpc/lib/web/plugs/supported_watcher_modes.ex]: 1aa0f3d

**Pull request merged**